File: vscode/src/utils/inlay/PubInlayUtils.ts
import * as vscode from "vscode";
import { I18nKey } from "../../common/I18nKey";
import { Result } from "../../common/Result";
import { GithubRepoInfoService } from "../service/GithubRepoInfoService";
import { PubPkgInfoService } from "../service/PubPkgInfoService";
import { I18n } from "../I18n";
import { ProgressUtils } from "../ProgressUtils";
import { UiUtils } from "../UiUtils";

export class PubInlayUtils {
  static addPubDepInlay(
    hints: vscode.InlayHint[],
    pkg: string,
    position: vscode.Position,
  ): void {
    const pubRes = PubPkgInfoService.getInstance().getPackageInfo(pkg);

    switch (pubRes.result) {
      case Result.NONE:
        if (
          PubPkgInfoService.getInstance().hasFailure(pkg) &&
          !PubPkgInfoService.getInstance().isRequestRunning(pkg)
        ) {
          UiUtils.addInlay(
            hints,
            position,
            I18n.message("failedPubMeta"),
            undefined,
            undefined,
            `pub:${pkg}`,
          );
          return;
        }

        void Promise.resolve(
          ProgressUtils.runBackground(`DepLens: Fetch pub ${pkg}`, () =>
            PubPkgInfoService.getInstance().fetchPackageInfo(pkg),
          ),
        ).catch((error: unknown) => {
          console.warn(`Failed to load pub info for ${pkg}`, error);
        });
        UiUtils.addInlay(
          hints,
          position,
          I18n.message("loadingPubMeta"),
          undefined,
          undefined,
          `pub:${pkg}`,
        );
        return;

      case Result.PENDING:
        UiUtils.addInlay(
          hints,
          position,
          I18n.message("loadingPubMeta"),
          undefined,
          undefined,
          `pub:${pkg}`,
        );
        return;

      case Result.SUCCESS:
        break;

      case Result.FAILURE:
        UiUtils.addInlay(
          hints,
          position,
          I18n.message("failedPubMeta"),
          undefined,
          undefined,
          `pub:${pkg}`,
        );
        return;
    }

    const pubInfo = pubRes.data;
    if (!pubInfo) {
      UiUtils.addInlay(
        hints,
        position,
        I18n.message("failedPubMeta"),
        undefined,
        undefined,
        `pub:${pkg}`,
      );
      return;
    }

    const url = pubInfo.githubUrl;
    if (!url) {
      UiUtils.addInlay(
        hints,
        position,
        I18n.message(I18nKey.noGithubUrl),
        undefined,
        undefined,
        `pub:${pkg}`,
      );
      return;
    }

    const repoKey = GithubRepoInfoService.getRepoKey(url);
    if (!repoKey) {
      UiUtils.addInlay(
        hints,
        position,
        I18n.message(I18nKey.invalidGithubUrl),
        undefined,
        undefined,
        `pub:${pkg}`,
      );
      return;
    }

    const repoRes = GithubRepoInfoService.getInstance().getRepoInfo(
      repoKey.owner,
      repoKey.repo,
    );
    switch (repoRes.result) {
      case Result.NONE:
        void Promise.resolve(
          ProgressUtils.runBackground(
            `DepLens: Fetch GitHub ${repoKey.owner}/${repoKey.repo}`,
            () =>
              GithubRepoInfoService.getInstance().fetchRepoInfo(
                repoKey.owner,
                repoKey.repo,
              ),
          ),
        ).catch((error: unknown) => {
          console.warn(
            `Failed to load repo info for ${repoKey.owner}/${repoKey.repo}`,
            error,
          );
        });
        UiUtils.addInlay(
          hints,
          position,
          I18n.message(I18nKey.loadingGithub),
          undefined,
          `https://github.com/${repoKey.owner}/${repoKey.repo}`,
          `github:${repoKey.owner}/${repoKey.repo}`,
        );
        return;

      case Result.PENDING:
        UiUtils.addInlay(
          hints,
          position,
          I18n.message(I18nKey.loadingGithub),
          undefined,
          `https://github.com/${repoKey.owner}/${repoKey.repo}`,
          `github:${repoKey.owner}/${repoKey.repo}`,
        );
        return;

      case Result.SUCCESS: {
        const stars = repoRes.data?.stars ?? "0";
        const updated = repoRes.data?.updatedDate ?? "N/A";
        UiUtils.addInlay(
          hints,
          position,
          `⭐ ${stars} • ${I18n.message(I18nKey.lastUpdated)} ${updated}`,
          undefined,
          `https://github.com/${repoKey.owner}/${repoKey.repo}`,
          `github:${repoKey.owner}/${repoKey.repo}`,
        );
        return;
      }

      case Result.FAILURE:
        UiUtils.addInlay(
          hints,
          position,
          I18n.message(I18nKey.failedGithub),
          undefined,
          `https://github.com/${repoKey.owner}/${repoKey.repo}`,
          `github:${repoKey.owner}/${repoKey.repo}`,
        );
        return;
    }
  }
}

File: jetbrains/src/main/kotlin/deplens/utils/inlay/PubInlayUtils.kt
package deplens.utils.inlay

import com.intellij.codeInsight.hints.declarative.InlayTreeSink
import com.intellij.openapi.diagnostic.logger
import com.intellij.psi.PsiFile
import deplens.common.I18nKey
import deplens.common.Result
import deplens.utils.I18n
import deplens.utils.ProgressUtils
import deplens.utils.UiUtils
import deplens.utils.service.GithubRepoInfoService
import deplens.utils.service.PubPkgInfoService

object PubInlayUtils {

    private val LOG = logger<PubInlayUtils>()

    fun addPubDepInlay(file: PsiFile, sink: InlayTreeSink, pkg: String, offset: Int) {
        val project = file.project
        val vFile = file.virtualFile
        val pubRes = PubPkgInfoService.getPackageInfo(pkg)

        when (pubRes.result) {
            Result.NONE -> {
                if (PubPkgInfoService.hasFailure(pkg) && !PubPkgInfoService.isRequestRunning(pkg)) {
                    UiUtils.addInlay(
                        sink,
                        offset,
                        I18n.message("failedPubMeta"),
                        retryToken = "pub:$pkg"
                    )
                    return
                }
                ProgressUtils.runBackground(file.project, "DepLens: Fetch pub $pkg") {
                    try {
                        PubPkgInfoService.fetchPackageInfo(pkg) {
                            var dispatchedGithubFetch = false
                            val refreshedPubRes = PubPkgInfoService.getPackageInfo(pkg)

                            if (refreshedPubRes.result == Result.SUCCESS) {
                                val url = refreshedPubRes.data?.githubUrl.orEmpty()
                                val repoKey = GithubRepoInfoService.getRepoKey(url)
                                if (repoKey != null) {
                                    val repoRes = GithubRepoInfoService.getRepoInfo(repoKey.owner, repoKey.repo)
                                    if (repoRes.result == Result.NONE) {
                                        dispatchedGithubFetch = true
                                        ProgressUtils.runBackground(
                                            file.project,
                                            "DepLens: Fetch GitHub ${repoKey.owner}/${repoKey.repo}"
                                        ) {
                                            GithubRepoInfoService.fetchRepoInfo(repoKey.owner, repoKey.repo) {
                                                if (vFile != null) {
                                                    UiUtils.refreshInlayHints(project, vFile)
                                                }
                                            }
                                        }
                                    }
                                } else if (url.isNotBlank()) {
                                    LOG.warn("Failed to get repo key for $pkg")
                                }
                            }

                            if (!dispatchedGithubFetch && vFile != null) {
                                UiUtils.refreshInlayHints(project, vFile)
                            }
                        }
                    } catch (e: Exception) {
                        LOG.warn("Failed to load pub info for $pkg", e)
                    }
                }
                UiUtils.addInlay(
                    sink,
                    offset,
                    I18n.message("loadingPubMeta"),
                    retryToken = "pub:$pkg"
                )
                return
            }

            Result.SUCCESS -> {}

            Result.PENDING -> {
                UiUtils.addInlay(
                    sink,
                    offset,
                    I18n.message("loadingPubMeta"),
                    retryToken = "pub:$pkg"
                )
                return
            }

            else -> {
                UiUtils.addInlay(
                    sink,
                    offset,
                    I18n.message("failedPubMeta"),
                    retryToken = "pub:$pkg"
                )
                return
            }
        }

        val pubInfo = pubRes.data ?: run {
            UiUtils.addInlay(
                sink,
                offset,
                I18n.message("failedPubMeta"),
                retryToken = "pub:$pkg"
            )
            return
        }

        val url = pubInfo.githubUrl
        if (url.isNullOrBlank()) {
            UiUtils.addInlay(
                sink,
                offset,
                I18n.message(I18nKey.noGithubUrl),
                retryToken = "pub:$pkg"
            )
            return
        }

        val repoKey = GithubRepoInfoService.getRepoKey(url)
        if (repoKey == null) {
            UiUtils.addInlay(
                sink,
                offset,
                I18n.message(I18nKey.invalidGithubUrl),
                retryToken = "pub:$pkg"
            )
            return
        }

        val repoRes = GithubRepoInfoService.getRepoInfo(repoKey.owner, repoKey.repo)
        val displayText = when (repoRes.result) {
            Result.NONE -> {
                ProgressUtils.runBackground(
                    file.project,
                    "DepLens: Fetch GitHub ${repoKey.owner}/${repoKey.repo}"
                ) {
                    GithubRepoInfoService.fetchRepoInfo(repoKey.owner, repoKey.repo) {
                        if (vFile != null) {
                            UiUtils.refreshInlayHints(project, vFile)
                        }
                    }
                }
                I18n.message(I18nKey.loadingGithub)
            }

            Result.PENDING -> I18n.message(I18nKey.loadingGithub)
            Result.SUCCESS -> "⭐ ${repoRes.data?.stars ?: 0} • ${I18n.message(I18nKey.lastUpdated)} ${repoRes.data?.updatedDate ?: "N/A"}"
            else -> I18n.message(I18nKey.failedGithub)
        }

        val githubUrl = "https://github.com/${repoKey.owner}/${repoKey.repo}"
        UiUtils.addInlay(
            sink,
            offset,
            displayText,
            githubUrl = githubUrl,
            retryToken = "github:${repoKey.owner}/${repoKey.repo}"
        )
    }
}
